Add ADR for TAD tool development (#104)
MinBZK/ai-validation: docs/ADRs/012-tad-tool.md

# ADR-0012 Tool for Transparency of Algorithmic Decision making

## Context

The Dutch government demands that government organisations are transparent about the use of
high impact algorithms. This is a requirement of
the [Open Government Act](https://www.rijksoverheid.nl/onderwerpen/wet-open-overheid-woo).
This transparency can start with registering an algorithm in the
[Algoritmeregister](https://algoritmes.overheid.nl/).

> Transparency and validation can be achieved by doing assessments,
> like [IAMA](https://www.rijksoverheid.nl/documenten/rapporten/2021/02/25/impact-assessment-mensenrechten-en-algoritmes)
> for decision making and technical evaluations to give insight in the features and outcomes of an algorithm. These
> outcomes can be combined to create a report that can be shared.

Reviewer: ```suggestion
Validation can be achieved by proper evaluation of algorithmic performance, bias and fairness and by doing assessments (like [IAMA](https://www.rijksoverheid.nl/documenten/rapporten/2021/02/25/impact-assessment-mensenrechten-en-algoritmes)).  Transparency can be achieved by sharing these outcomes in a standardized way. 
```

> There are several tools available on the market. Some support only assessments, others already

Reviewer: ```suggestion
There are several existing open source tools. Some support only assessments, others already
```

Moreover, maybe also include our shortlist so we can be as specific as possible.

Author: I have made the suggested change and added the link.

combine more features and can generate a report. There is however no tool that supports

> all [requirements](https://minbzk.github.io/ai-validation/Projects/TAD/Tools/) we have.

Reviewer: ```suggestion
all the [requirements](https://minbzk.github.io/ai-validation/Projects/TAD/Tools/) we have.
```

## Assumptions

> * Collaborating or extending another project will not give us the tool we are looking for.

Author: Yes, will do, when I originally wrote this that branch was not merged yet into main.

> * Collaborating or extending another project will not give us the tool we are looking for.

Author: I have added a link to the comparison earlier in the document.

* We can reuse certain components, like the
[plugins from AIVerify](https://github.com/IMDA-BTG/aiverify/tree/main/stock-plugins) for technical tests.

> * The tool will use a loose modules design, by scanning directories, to gather plugin functionality.

Reviewer: I think this sentence needs to be split up into multiple sentences or item, because I am not sure how I should read this now. 

I also would use another word than "loose modules design". Maybe something like "loosely coupled module design" or "The tool will use a design based on loose modules."

Author: I reworded the sentence.

# ADR-0012 Tool for Transparency of Algorithmic Decision making

## Context

The Dutch government demands that government organisations are transparent about the use of
high impact algorithms. This is a requirement of
the [Open Government Act](https://www.rijksoverheid.nl/onderwerpen/wet-open-overheid-woo).
This transparency can start with registering an algorithm in the
[Algoritmeregister](https://algoritmes.overheid.nl/).

Transparency and validation can be achieved by doing assessments,
like [IAMA](https://www.rijksoverheid.nl/documenten/rapporten/2021/02/25/impact-assessment-mensenrechten-en-algoritmes)
for decision making and technical evaluations to give insight in the features and outcomes of an algorithm. These
outcomes can be combined to create a report that can be shared.

There are several tools available on the market. Some support only assessments, others already
combine more features and can generate a report. There is however no tool that supports

> all [requirements](https://minbzk.github.io/ai-validation/Projects/TAD/Tools/) we have.

Author: I have thought about mentioning that, however I'm not sure if it is relevant or where it should appear. It also is not a real relevant issue for developing our own TAD, as we very well could put this functionality in f.e. AIVerify. What do others think?

> all [requirements](https://minbzk.github.io/ai-validation/Projects/TAD/Tools/) we have.

Reviewer 2: Yeah good point, maybe it is more relevent for the TAD project compared to the TAD tool, right now I have incorporated this in the PR for the project description https://github.com/MinBZK/ai-validation/pull/111

> all [requirements](https://minbzk.github.io/ai-validation/Projects/TAD/Tools/) we have.

Author: Yes, it should. Actually it feels like it should point to a 'landing page' on our exisiting tools section, which we don't have yet. I will add one and update the link.

> all [requirements](https://minbzk.github.io/ai-validation/Projects/TAD/Tools/) we have.

Author: Resolved this discussion as this link is fixed in another conversation.

## Assumptions

* Collaborating or extending another project will not give us the tool we are looking for.

> * We can reuse certain components, like the plugins from AI-Verify for technical tests.

Reviewer 2: ```suggestion
* We can reuse certain components, like the plugins from AIVerify for technical tests.
```

* The tool will use a loose modules design, by scanning directories, to gather plugin functionality.
* Plugins will have to implement an Interface to be picked up by the system.
* We can, to some extend, re-use the already done POCs or findings from these POCs.

## Decision

We will develop our own tool. Where possible this tool should be able to re-use certain components of other open-source
tools.

## Risks

* It is a lot of work to create our own tool.
* We may not have all the knowledge yet to create technical tests.

## Consequences

We will have a tool that is tailored to our needs and we won't have to discuss the direction we take and choices we make
with another party.

> ## More Information

Reviewer 2: I would remove this if it doesn't have any content

# ADR-0012 Tool for Transparency of Algorithmic Decision making

## Context

We are considering tooling for organisations to get more grip on their algorithms. Tooling for, for instance bias and fairness tests, and assessments (like [IAMA](https://www.rijksoverheid.nl/documenten/rapporten/2021/02/25/impact-assessment-mensenrechten-en-algoritmes)).

Transparency, we think, can be fostered by sharing reports from such a tool in a standardized way.

There are several existing open source

> [tools](https://minbzk.github.io/ai-validation/Projects/TAD/Existing%20Tools/Comparison/tools/) which
> we [have assessed](https://minbzk.github.io/ai-validation/Projects/TAD/Existing%20Tools/Comparison/tools_comparison/).

Reviewer 3: ```suggestion
[tools](../Projects/TAD/Existing%20Tools/Comparison/tools/) which
we [have assessed](../Projects/TAD/Existing%20Tools/Comparison/tools_comparison/).
```

We should really fix that url to have a `-` instead of a space though

Some support only assessments, others already combine more features and can generate a report.
There is however no tool

> that supports all the [requirements](https://minbzk.github.io/ai-validation/Projects/TAD/Tools/) we have.

Author: ```suggestion
that supports all the [requirements](https://minbzk.github.io/ai-validation/Projects/TAD/Existing Tools/Comparison/requirements/) we have.
```

> that supports all the [requirements](https://minbzk.github.io/ai-validation/Projects/TAD/Existing Tools/Comparison/requirements/) we have.

Reviewer 3: ```suggestion
that supports all the [requirements](../Projects/TAD/Existing Tools/Comparison/requirements/) we have.
```

Author: Oh great, it resolves.. check, will remember for next time.

> # ADR-0012 Tool for Transparency of Algorithmic Decision making

Reviewer 3: We keep project ADRs in Projects (which, in this case, happens to be in the same repo, on a different path).

Not sure we need an ADR to decided on that?

Customary, every new series of ADRs starts with "an ADR on the fact that we use ADRs".

> # ADR-0012 Tool for Transparency of Algorithmic Decision making

Author: "We keep project ADRs in Projects" sounds like a decision/WoW. An ADR is a bit much, but maybe in Way-of-Working we can keep a list of such decisions?

Also fine with using an ADR to declare the ADR, point is, the file is duplicated instead of f.e. referenced which could (even if unlikely) lead to different files.. so I wondered if referencing it may be a solution to that issue?

> # ADR-0012 Tool for Transparency of Algorithmic Decision making

Reviewer 3: If you think it warrants and ADR or note somewhere: we should to that. 

For the second: I don't think it matters tbh.

## Context

> The Government of the Netherlands demands that government organisations are transparent about the use of
> high impact algorithms. This is a requirement of
> the [Open Government Act](https://www.rijksoverheid.nl/onderwerpen/wet-open-overheid-woo).
> This transparency can start with registering an algorithm in the
> [Algoritmeregister](https://algoritmes.overheid.nl/).
> 
> Validation can be achieved by proper evaluation of algorithmic performance,
> bias and fairness and by doing assessments (like [IAMA](https://www.rijksoverheid.nl/documenten/rapporten/2021/02/25/impact-assessment-mensenrechten-en-algoritmes)).
> Transparency can be achieved by sharing these outcomes in a standardized way.

Reviewer 3: I'm not sure we should start with this, this is rather the context for our team, I think we can zoom in for this tool. 

```suggestion
We are considering tooling for organisation to get more grip on their algorithms. Tooling for, for instance bias and fairness tests and assessments (like [IAMA](https://www.rijksoverheid.nl/documenten/rapporten/2021/02/25/impact-assessment-mensenrechten-en-algoritmes)).

Transparency, we think, can be fostered by sharing reports from such a tool in a standardized way.
```

There are several existing open source
[tools](https://minbzk.github.io/ai-validation/Projects/TAD/Existing%20Tools/Comparison/tools/) which
we [have assessed](https://minbzk.github.io/ai-validation/Projects/TAD/Existing%20Tools/Comparison/tools_comparison/).
Some support only assessments, others already combine more features and can generate a report.
There is however no tool
that supports all the [requirements](https://minbzk.github.io/ai-validation/Projects/TAD/Tools/) we have.

These are our main requirements:

* The tool offers A one-stop shop for all aspects of the project for all stakeholders for all tasks.

> * The tool supports a workflow to track different stages of the project.

Reviewer 3: ```suggestion
* The tool supports a workflow to track different stages of the project, it should support lifecycle management.
```

> * The tool can run different technical tests on a model.

Reviewer 3: ```suggestion
* The tool can run many technical tests on a model.
```

> * The tool supports filling out assesments.

Author: ```suggestion
* The tool supports filling out assessments.
```

> * The tool supports discussion features.

Reviewer 3: ```suggestion
* The tool supports capturing discussion around technical tests and assessments.
```

* The tool offers UI access to the above requirements.
* The tool can save results to a system card (or cards supported by system cards, like model, metrics and assessment).

> * The tool can push the different cards to a VCS like GitHub to allow for an audit trail.

Reviewer 3: ```suggestion
* The tool can commit the different cards to a VCS, such as git, to allow for an audit trail.
```

## Assumptions

* Collaborating or extending another project will not give us the tool we are looking for.

> * We can reuse certain components, like the
> [plugins from AIVerify](https://github.com/IMDA-BTG/aiverify/tree/main/stock-plugins) for technical tests.

Reviewer 3: ```suggestion
* We can reuse certain components, like the
[plugins from AIVerify](https://github.com/IMDA-BTG/aiverify/tree/main/stock-plugins), or existing libraries, for technical tests.
```

* The tool will use a design based on loose coupled modules. This can be done by scanning directories,
  to gather modules that implement a certain Interface.
* Plugins will have to implement an Interface to be picked up by the system.
* We can, to some extend, re-use the already existing POCs or findings from these POCs.

## Decision

> We will develop our own tool. Where possible this tool should be able to re-use certain components of other open-source

Reviewer 2: Wouldn't we maybe want to make a set of own tools? Maybe one tool that does actually the workflow managament, and another that does the more advanced technical tests?

Author: I'm not sure yet if it will be one thing or more or a tool set.. what if we make it a bit more 'general', does that work for you @laurensWe ?

```suggestion
We will build our own solution. Where possible this solution should be able to re-use certain components of other open-source
```

Author: no it is not, I guess it is marked outdated due to... I'm not sure why..

Author: I have added the change in another commit.

tools.

## Risks

* It is a lot of work to create our own tool.

> * We may not have all the knowledge yet to create technical tests.

Reviewer 3: I don't think we should create any technical test, we should reuse existing tests (we already mention this in our assumptions)

Reviewer 3: I believe a bigger risk is that we may struggle to get uptake of the tool if we are not aligned with envisioned users of the tool

> ## Consequences
> 
> We will have a tool that is tailored to our needs and we do not have to discuss the direction we take and choices we make
> with third parties outside our organization.

Reviewer 3: I would leave this out (plus I disagree :) )
```suggestion
```

Author: Would you care to elaborate why you disagree (here or in a meeting).. I'm curious :) While re-reading, again the wording may be off, so I wonder if this works better.. 

```suggestion
## Consequences

We can develop a solution that is tailored to the needs of our users and we can involve them to discuss the direction we take and choices we make
```

Reviewer 3: Of course! 

I think we should aim to build a tool tailored to the needs of our stakeholders and it will involved lots of discussions with many, also outside of our organization.

Author: Thanks. I have edited the consequence to reflect that. Is it ok to put it like this or would you still rather leave the whole block out?
